Sources/HAP/Server/Device.swift

```diff
@@ -144,15 +230,107 @@ public class Device {
         // Accessory Server.
         //
-        // TODO: SwiftFoundation in Swift 4.0 cannot encode/decode UInt64,
-        // which is the data-type we wanted to use here. We can change it back
-        // to UInt64 once the following commit has made it into a release:
-        // https://github.com/apple/swift-corelibs-foundation/commit/64b67c91479390776c43a96bd31e4e85f106d5e1
-        var idGenerator = (1...InstanceID.max).makeIterator()
-        for accessory in accessories {
+
+        // Obtain new aid's for any accessories which don't already have one
+
+        self.accessories = [Accessory]()
+
+        // The first accessory must be aid 1
+        accessories[0].aid = 1
+        
+        addAccessories(accessories)
+    }
+    
+    public func addAccessories(_ newAccessories: [Accessory]) {
+ 
+        var configurationUpdated = false
+        
+        accessories.append(contentsOf: newAccessories)
+        
+        for accessory in newAccessories {
             accessory.device = self
-            accessory.aid = idGenerator.next()!
+            if (accessory.aid == 0) {
+                
+                // Check to see if the aid has been stored in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                if let aid = configuration.aidForAccessoryUniqueId[dsuid] {
+                    accessory.aid = aid
+                }
+            }
         }
+
+        for accessory in newAccessories {
+            // Verify that the aid is indeed unique
+            if accessory.aid != 0,
+               !isUniqueAID(accessory.aid, ignoring: accessory) {
+                logger.info("Accessory \(accessory.info.name.value ?? "unknown") has a duplicate accessory ID \(accessory.aid), replacing with fresh IID")
+                accessory.aid = 0
+            }
+            if (accessory.aid == 0) {
+                // Obtain a new aid, which has not already been used
+                repeat {
+                    accessory.aid = configuration.nextAID
+                } while (!isUniqueAID(accessory.aid, ignoring: accessory))
+                configurationUpdated = true
+
+                // Store the aid in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                configuration.aidForAccessoryUniqueId[dsuid] = accessory.aid
+            }
+        }
+        
+        // write configuration data to persist updated aid's and notify listeners
+        
+        if configurationUpdated {
+            updatedConfiguration()
+        }
+
+        
+    }
+    
+    public func removeAccessories(_ unwantedAccessories: [Accessory]) {
+        
+        var configurationUpdated = false
+        
+        for accessory in unwantedAccessories {
+            // Ensure the initial accessory is not removed, and that the accessory is in the list
+            if accessory.aid != 1 {
```

Comment:
Why shouldn't we remove the initial accessory? Also; we shouldn't silently ignore that, but trap as it is invalid API usage. Thus use something like `preconditionFailure()`.

Reply:
According to the spec "2.5.3.3 Primary HAP Accessory Object", the first accessory exposed by a Bridge has to be the bridge itself, so cannot be removed. The called should create a new Device instead. I've modified the func to use precondition()